File: Month-3/Week-09/day-63/c++/main.cpp
/**
 * DAY-63 C++ version
 * 
 * Just run 'make' to compile and run all the tests
 */
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "../../../../dependencies/c++/doctest.h" //https://github.com/onqtam/doctest

/**
 * The rabbit problem. How long it will take for them to outnumber humans (target)?
 * 
 * Every month a fertile female will have 14 offspring (5 males and 9 females).
 * A female rabbit is fertile when it has reached the age of 4 months, they never stop being fertile.
 * Rabbits die at the age of 8 years (96 months).
 * 
 * @param Number of 2-month-old male rabbits to start
 * @param Number of 2-month-old female rabbits to start
 * @return Number of months to hit the target
 */ 
int rabbit(int male, int female, long int target) {           
    const int ERROR_CODE = -1;    
    const int INITIAL_BATCH_AGE = 2; //each starting rabbit is 2-month-old
    const int AGES = 96; //the lifespan of a rabbit (they die at 96 months)
    const int AGE_ALIVE_LIMIT = 95;
    const int AGE_FERTILE_BEGIN = 4; //A female rabbit is fertile when it has reached the age of 4 months, they never stop being fertile.
    const int NEW_MALES_RATE = 5;
    const int NEW_FEMALES_RATE = 9;

    //check invalid input
    if (female < 0 || male < 0) {
        return -1; //invalid input
    }
    
    //without an starting female the population will not grow
    if ((male + female) < target && female == 0 ) {
        return ERROR_CODE;
    }
        
    int month = 0; //starting month count
    long int f[AGES] = {0}; //initialize all positions with zero for females
    long int m[AGES] = {0}; //initialize all positions with zero for males
        
    f[INITIAL_BATCH_AGE] = female; //input some initial 2-month-old female rabbits
    m[INITIAL_BATCH_AGE] = male; ////input some initial 2-month-old female rabbits
    
    while (true) {
        long int aliveRabbits = 0; //count the alive rabbits on this month

        for (int i = AGE_ALIVE_LIMIT; i >= 0 && aliveRabbits < target; i--) {
            
            //Offspring production
            long int newBornFemale = 0;
            long int newBornMale = 0;
            if (i > AGE_FERTILE_BEGIN) { //A fertile female will always produce 14 offspring (5 male, 9 female)
                newBornFemale = f[i] * NEW_FEMALES_RATE; 
                newBornMale = f[i] * NEW_MALES_RATE;
            } 

            aliveRabbits += f[i] + m[i];

            //Age rabbits
            const int nextGen = i + 1;
            f[nextGen] = f[i];
            m[nextGen] = m[i];                    
            f[i] = 0;
            m[i] = 0;

            //the first generation (index 0) will always be the newborns
            f[0] += newBornFemale;
            m[0] += newBornMale;                        
        }
                    
        if (aliveRabbits >= target) { //check if the rabbit population hit the target            
            return month;
        }    
        month++;
    }    
}


/**
 * Tests
 */
TEST_CASE("Tests")
{  
   CHECK(rabbit(2, 4, 6) == 0);
   CHECK(rabbit(2, 4, 1000000000) == 32);
   CHECK(rabbit(2, 4, 15000000000) == 36);
   CHECK(rabbit(5, 0, 6) == -1);
   CHECK(rabbit(-1, 2, 6) == -1);
}

File: Month-3/Week-09/day-63/Go/main.go
package main

import "fmt"

func main() {
	input1 := []float64{2, 4, 1000000000}
	input2 := []float64{2, 4, 15000000000}
	input3 := []float64{2, 4, 184000000000000000000000000}

	fmt.Println(rabbitSwarm(input1))
	fmt.Println(rabbitSwarm(input2))
	fmt.Println(rabbitSwarm(input3))
}

type rabbit struct {
	male, female, all float64
}

func rabbitSwarm(input []float64) (months, dead float64) {
	initialRabbit := rabbit{input[0], input[1], input[0] + input[1]}
	var arrRabbits [96]rabbit
	var allRabbits = arrRabbits[:]
	allRabbits[1] = initialRabbit
	total := initialRabbit.all
	dead = 0
	for months = 0; total < input[2]; months++ {
		var fertile = allRabbits[4:]
		newRabbits := rabbit{}
		for i := 0; i < len(fertile); i++ {
			newRabbits.female += fertile[i].female * 9
			newRabbits.male += fertile[i].female * 5
		}
		newRabbits.all += newRabbits.female + newRabbits.male
		//Remove dead rabbits
		total -= allRabbits[95].all
		dead += allRabbits[95].all
		//Add new rabbits
		total += newRabbits.all
		// Remove the last element from the array (Rabbits have died)
		allRabbits = allRabbits[:len(allRabbits)-1]
		// Insert the newest element to the front of the array (Rabbits are born)
		allRabbits = append([]rabbit{newRabbits}, allRabbits...)
	}
	return months - 1, dead
}
